server/queue/queue-pulsar/src/main/scala/org/apache/james/queue/pulsar/PulsarMailQueue.scala

```diff
@@ -0,0 +1,600 @@
+    } yield null
+  }
+
+  /**
+   * @inheritdoc
+   */
+  override def deQueue(): Publisher[MailQueue.MailQueueItem] = dequeuePublisher
+
+  /**
+   * @inheritdoc
+   */
+  override def close(): Unit = {
+    enqueue.complete()
+    requeue.complete()
+    dequeueControl.stop()
+    scheduledConsumerControl.stop()
+    filterFlowControl.stop()
+    client.close()
+    system.terminate() //FIXME: to remove when injected
+  }
+
+  /**
+   * @inheritdoc
+   */
+  override def getSize: Long = getSize(outTopic, subscription) + getSize(scheduledTopic, scheduledSubscription)
+
+
+  private def getSize(topic: Topic, subscription: Subscription): Long = {
+    try {
+      val subscriptions = admin.topics().getStats(topic.name).getSubscriptions
+      val maybeStats = Option(subscriptions.get(subscription.name))
+      maybeStats.map(_.getMsgBacklog).getOrElse(0)
+    } catch {
+      case _: NotFoundException => 0L
+    }
+  }
+
+  /**
+   * @inheritdoc
+   */
+  override def flush(): Long = {
+    def lastScheduledMessageId(f: MessageId => Long): Long = lastMessage(scheduledTopic)
+      .map(_.messageId)
+      .fold(0L)(f)
+
+    lastScheduledMessageId { lastMessageId =>
+      val flushStart = Instant.now()
+
+      def isScheduledAfterFlush(message: ConsumerMessage[String]) = Instant.ofEpochMilli(message.eventTime.value).isAfter(flushStart)
+
+      def putMessageInOutTopic(message: ConsumerMessage[String]) = requeue.offer(ProducerMessage(message.value))
+
+      //prevents normal deque flow of scheduled messages by moving the scheduled subscription to last know message
+      admin.topics().resetCursor(scheduledTopic.name, scheduledSubscription.name, lastMessageId)
+
+      val copy = read(
+        ReaderConfig(scheduledTopic, startMessage = Message(MessageId.earliest), startMessageIdInclusive = true)
+      )
+        .filter(isScheduledAfterFlush) //avoid duplicate delivery of messages which are already handled by scheduledSubscription
+        .filter(_.messageId.underlying <= lastMessageId.underlying) //stop copying message at lastMessageId
+        .runFold(0L) { (counter, message) =>
+          putMessageInOutTopic(message)
+          counter + 1
+        }
+      Await.result(copy, Duration.Inf)
+    }
+  }
+
+  private def read(config: ReaderConfig)(implicit executionContext: ExecutionContext): Source[ConsumerMessage[String], NotUsed] =
+    Source.unfoldResourceAsync[ConsumerMessage[String], Reader[String]](
+      create = () => {
+        Future.successful(
+          client.reader(
+            config = config
+          )
+        )
+      },
+      read = reader => {
+        if (reader.hasMessageAvailable) reader.nextAsync.map(Some(_))
+        else Future.successful(None)
+      },
+      close = reader => reader.closeAsync.map(_ => Done))
+
+
+  /**
+   * @inheritdoc
+   */
+  override def clear(): Long = {
+    val count = getSize()
+    admin.topics().delete(outTopic.name, true)
+    count
+  }
```

Comment:
Shouldn't clear also get rid of the scheduled topic?

After clearing, can messages still transit through the mailqueue? (I expect it should, but I do not see how this can work if we nuke the out topic... New producers would re-create automatically a missing topic?)

Reply:
> New producers would re-create automatically a missing topic?

yes, n pulsar there is no need to [explicitely create topics](https://pulsar.apache.org/docs/en/concepts-messaging/#topics) 

>Shouldn't clear also get rid of the scheduled topic?

This behavior is not defined by the existing contracts, I'll defer to @mbaechler on which behaviour makes the most sense.

Reply:
> This behavior is not defined by the existing contracts, I'll defer to @mbaechler on which behaviour makes the most sense.

It is a pity. IMO it should.